src/os/bluestore/BlockDevice.cc

```diff
@@ -82,10 +87,11 @@ BlockDevice *BlockDevice::create(CephContext* cct, const string& path,
     return new PMEMDevice(cct, cb, cbpriv);
   }
 #endif
-
+#if defined(HAVE_LIBAIO)
   if (type == "kernel") {
     return new KernelDevice(cct, cb, cbpriv);
   }
+#endif
```

Comment:
@tchaikov Eventually I was able to build with spdk [1] but this did not help for "unknown backend kernel" error. Which is quite understandable: "kernel" type is default and it is disabled when built without LIBAIO.

So it looks like blustore can only work on FreeBSD if it is built with spdk and nvme devices are used for osd?

[1] To build spdk on FreeBSD I had to add the following hacks:

1) in cmake/modules/BuildDPDK.cmake, change the hardcoded "make" to "gmake";
2) add GNU sed to my path (some script uses sed command "m" not supported by BSD sed);
3) in src/spdk/mk/spdk.common.mk, change `-march=native` compiler option to `-march=corei7` (to workaround "unknown target CPU 'pentium2'" error, observed in my env).

Reply:
@tchaikov @trociny 
http://cephdev.digiware.nl:8180/jenkins/job/ceph-master/1463/consoleFull contains the full log of importing the this PR. Looks like I also tried with `SPDK=OFF`, so that is a bummer. Second issue is  that does not enable the usage of BlueStore yet. So there is no tests with bluestore tested.
I guess that that is what @trociny does with a separate test, whereas I only tested the tests from QA in the current shape they are now.

And yes to some of the limitation from what I read on the SPDK page is that it requires NVME to actually build the devices on. So testing is going to be hard if you do not have NVMEs, which I do not.

But it is like @trociny, the problem is likely not due to this patch.
Running this patch again with `SPDK=ON`

Reply:
@tchaikov @trociny 
I was talking about the changes to `cmake/modules/BuildDPDK.cmake`.
There is a extern process which is called with `make showconfig`, that needs to be executed with the correct invocation of MAKE. I've tried `$(MAKE)`, `${MAKE}`  and '${CMAKE_MAKE_PROGRAM}'
And only the last one really did the work. Then i could build manually. But my Jenkins build is still upset.

Thus came up with the following diff:
```
index 1b824ae5b9..99e4ca1ac3 100644
--- a/cmake/modules/BuildDPDK.cmake
+++ b/cmake/modules/BuildDPDK.cmake
@@ -51,13 +51,14 @@ function(do_build_dpdk dpdk_dir)
   set(target "${arch}-${machine_tmpl}-${execenv}-${toolchain}")

   execute_process(
-    COMMAND make showconfigs
+    COMMAND ${CMAKE_MAKE_PROGRAM} showconfigs
     WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/src/spdk/dpdk
     OUTPUT_VARIABLE supported_targets
     OUTPUT_STRIP_TRAILING_WHITESPACE)
   string(REPLACE "\n" ";" supported_targets "${supported_targets}")
   list(FIND supported_targets ${target} found)
   if(found EQUAL -1)
+    message("  Platforms supported: " ${supported_targets})
     message(FATAL_ERROR "not able to build DPDK support: "
       "unsupported target \"${target}\"")
   endif()
@@ -65,8 +66,8 @@ function(do_build_dpdk dpdk_dir)
   include(ExternalProject)
   ExternalProject_Add(dpdk-ext
     SOURCE_DIR ${CMAKE_SOURCE_DIR}/src/spdk/dpdk
-    CONFIGURE_COMMAND $(MAKE) config O=${dpdk_dir} T=${target}
-    BUILD_COMMAND env CC=${CMAKE_C_COMPILER} $(MAKE) O=${dpdk_dir} EXTRA_CFLAGS=-fPIC
+    CONFIGURE_COMMAND ${CMAKE_MAKE_PROGRAM} config O=${dpdk_dir} T=${target}
+    BUILD_COMMAND env CC=${CMAKE_C_COMPILER} ${CMAKE_MAKE_PROGRAM} O=${dpdk_dir} EXTRA_CFLAGS=-fPIC
     BUILD_IN_SOURCE 1
     INSTALL_COMMAND "true")
   ExternalProject_Add_Step(dpdk-ext patch-config

```

Reply:
@wjwithagen oic. i incorporated the "make showconfigs" and "supported_targets" changes. but as i mentioned in last reply. we should stick with `$(MAKE)` whenever possible.

Reply:
@tchaikov 
Oke, we'll zee how that works out. If not we'll find out the hard way :)